pennylane/gradients/hadamard_gradient.py

```diff
@@ -335,28 +350,127 @@ def _hadamard_test(tape, trainable_param_idx, aux_wire) -> tuple[list, list]:
         new_batch.append(new_tape)
     return new_batch, sub_coeffs
 
+def _direct_hadamard_test(tape, trainable_param_idx, aux_wire) -> tuple[list, list]:
+
+    trainable_op, idx, _ = tape.get_operation(trainable_param_idx)
+
+    ops_to_trainable_op = tape.operations[: idx + 1]
+    ops_after_trainable_op = tape.operations[idx + 1 :]
+
+    # Get a generator and coefficients
+    sub_coeffs, generators = _get_pauli_generators(trainable_op)
+
+    measurements = tape.measurements
+
+    new_batch = []
+    new_coeffs = []
+    for idx, gen in enumerate(generators):
+        pos_rot = [qml.evolve(gen, np.pi/4)]
+        neg_rot = [qml.evolve(gen, -np.pi/4)]
+        pos_ops = ops_to_trainable_op + pos_rot + ops_after_trainable_op
+        neg_ops = ops_to_trainable_op + neg_rot + ops_after_trainable_op
+
+        pos_tape = qml.tape.QuantumScript(pos_ops, measurements, shots=tape.shots)
+        neg_tape = qml.tape.QuantumScript(neg_ops, measurements, shots=tape.shots)
+        new_batch.append(pos_tape)
+        new_batch.append(neg_tape)
+        new_coeffs.append(-1/2 * sub_coeffs[idx])
+        new_coeffs.append(1/2 * sub_coeffs[idx])
+    return new_batch, new_coeffs
+
+def _reversed_hadamard_test(tape, trainable_param_idx, aux_wire) -> tuple[list, list]:
+
```

Comment:
For reversed and direct reversed, can we add a check making sure there is only one meausrement, and raising a `ValueError` if not?

We could also add this validation up near the other validation checks, like the `assert_no_state`.

The benefit of having it here is that we've already dispatched by mode, but the benefit of having it by the other assertions is that we would only have to run it once.